Enable CryptoMiniSat-backed BV proofs

alex-ozdemir opened the pull request:
* Connect the plumbing so that BV proofs are enabled when using
CryptoMiniSat
* Also fixed a bug in CNF-proof generation
   * Specifically, CNF proofs broke when proving tautological clauses.
     Now they don't.

4tXJ7f commented on src/proof/cnf_proof.h:
```diff
@@ -164,6 +164,12 @@ class CnfProof {
+  // First field is true if the option is filled.
+  // Detects whether a clause has x v ~x for some x
+  // If so, returns the positive occurence's idx first, then the negative's
+  static std::tuple<bool, unsigned, unsigned> detectTrivialTautology(
```
We actually have a [`Maybe` type](https://github.com/CVC4/CVC4/blob/master/src/util/maybe.h) that you should be able to use for those purposes.

alex-ozdemir replied:
This is fantastic! Will do!

4tXJ7f commented on src/proof/bitvector_proof.cpp:
```diff
@@ -221,6 +221,14 @@ void BitVectorProof::printOwnedTerm(Expr term,
   }
 }
 
+void BitVectorProof::printEmptyClauseProof(std::ostream& os,
```
Remove the `= 0` from the header file.

alex-ozdemir replied:
Whoops! Will do

alex-ozdemir commented:
Thanks for the review, @4tXJ7f. I think I've addressed all of your comments. Let me know if I missed anything.

4tXJ7f commented on src/proof/cnf_proof.cpp:
```diff
@@ -431,6 +465,31 @@ void LFSCCnfProof::printCnfProofForClause(ClauseId id,
+    os << "(contra _ " << proofOfPos << " " << proofOfNeg << ")";
+  }
+  else
+  {
```
The indentation of the code within those parentheses is off.

alex-ozdemir replied:
Yeah. I'll format this file.

4tXJ7f replied:
Hmm, formatting the whole file is a bit excessive and makes it hard to see your changes. Any chance that you could just format your changes? [These](https://github.com/CVC4/CVC4/wiki/Developer-Guide#formatting-with-clang-format) explanations may be helpful. The easiest way is probably to do a `git reset HEAD~1` to remove your last commit, then add the non-formatting changes to a separate commit with `git add -p`, commit those changes, discard the remaining changes, then something like `git clang-format master`.

alex-ozdemir replied:
You're right -- thanks for keeping me honest...

I'll do it right and force-push over this.

4tXJ7f commented on src/proof/cnf_proof.cpp:
```diff
@@ -349,6 +349,40 @@ void CnfProof::collectAssertionsForClauses(const IdToSatClause& clauses,
 
+// Detects whether a clause has x v ~x for some x
+// If so, returns the positive occurence's idx first, then the negative's
+Maybe<std::pair<unsigned, unsigned>> CnfProof::detectTrivialTautology(
```
Use `size_t` for indices. `unsigned` is likely to be only 32-bit, even on 64-bit platforms.

alex-ozdemir replied:
:+1:

4tXJ7f commented on src/proof/theory_proof.cpp:
```diff
@@ -85,8 +85,29 @@ void TheoryProofEngine::registerTheory(theory::Theory* th) {
+          switch (options::bvProofFormat())
+          {
+            case theory::bv::BvProofFormat::BITVECTOR_PROOF_DRAT:
+              {
```
Clang-Format should actually not add an extra indentation here. Can you reformat?

alex-ozdemir replied:
:+1:

4tXJ7f commented on src/proof/cnf_proof.cpp:
```diff
@@ -431,6 +465,31 @@ void LFSCCnfProof::printCnfProofForClause(ClauseId id,
+    unsigned posIndexInClause = isTrivialTaut.value().first;
+    unsigned negIndexInClause = isTrivialTaut.value().second;
+    Trace("cnf-pf") << "; Indices " << posIndexInClause << " (+) and "
+                    << negIndexInClause << " (-) make this clause a taut"
```
```suggestion
                    << negIndexInClause << " (-) make this clause a tautology"
```

4tXJ7f commented on src/proof/cnf_proof.cpp:
```diff
@@ -349,6 +349,40 @@ void CnfProof::collectAssertionsForClauses(const IdToSatClause& clauses,
+    const prop::SatClause& clause)
+{
+  // a map from a SatVariable to its previous occurence's polarity and location
+  std::map<prop::SatVariable, std::pair<bool, unsigned>> varsToPolsAndIndices;
```
It probably doesn't change much in this case but I tend to prefer using an `unordered_map` if the keys don't need to be sorted (`unordered_map` is usually implemented as a hash map and `map` as some sort of tree data structure).

alex-ozdemir replied:
:+1:

4tXJ7f commented on src/proof/clausal_bitvector_proof.h:
```diff
@@ -87,6 +85,49 @@ class LfscClausalBitVectorProof : public ClausalBitVectorProof
+};
+
+/**
+ * An Extended Resolution proof for a bitvector problem
```
```suggestion
 * An Extended Resolution proof for a bit-vector problem
```

4tXJ7f commented on src/proof/clausal_bitvector_proof.h:
```diff
@@ -87,6 +85,49 @@ class LfscClausalBitVectorProof : public ClausalBitVectorProof
+};
+
+/**
+ * An LRAT proof for a bitvector problem
```
```suggestion
 * An LRAT proof for a bit-vector problem
```

4tXJ7f commented on src/proof/clausal_bitvector_proof.h:
```diff
@@ -87,6 +85,49 @@ class LfscClausalBitVectorProof : public ClausalBitVectorProof
+};
+
+/**
+ * A DRAT proof for a bitvector problem
```
```suggestion
 * A DRAT proof for a bit-vector problem
```

4tXJ7f commented on src/options/options_handler.cpp:
```diff
@@ -1769,7 +1769,8 @@ void OptionsHandler::proofEnabledBuild(std::string option, bool value)
+      && options::bvSatSolver() != theory::bv::SAT_SOLVER_CRYPTOMINISAT)
   {
     throw OptionException(
         "Eager BV proofs only supported when minisat is used");
```
```suggestion
        "Eager BV proofs only supported when MiniSat or CryptoMiniSat is used");
```

4tXJ7f commented on src/proof/clausal_bitvector_proof.cpp:
```diff
@@ -84,7 +87,16 @@ void ClausalBitVectorProof::calculateAtomsInBitblastingProof()
   }
-  Unimplemented();
+
+  // Empty any old record of which atoms for used
```
```suggestion
  // Empty any old record of which atoms were used
```

4tXJ7f commented on src/proof/clausal_bitvector_proof.cpp:
```diff
@@ -101,13 +113,91 @@ void LfscClausalBitVectorProof::printBBDeclarationAndCnf(std::ostream& os,
+{
+  Assert(options::bitblastMode() == theory::bv::BITBLAST_MODE_EAGER,
+         "the BV theory should only be proving bottom directly in the eager "
+         "bitblasting mode");
```
Instead of repeating these checks, we could just call the implementation in the base class. I am a bit on the fence on whether that would be cleaner or not but it is worth considering.

alex-ozdemir replied:
I thought about that too. I thought it would be weird to call a function named `print` in another `print` function, so I didn't.

4tXJ7f commented:
For the regressions, I think that you can actually just have a single test with the following header:

```
; REQUIRES: cryptominisat
; COMMAND-LINE: --bv-proof-format= lrat --check-proofs
; COMMAND-LINE: --bv-proof-format= drat --check-proofs
; COMMAND-LINE: --bv-proof-format= er --check-proofs
```

This should run the test three times with the different formats.

alex-ozdemir commented:
Okay, @4tXJ7f . I believe that I've addressed your comments. Let me know if this is not the case!

4tXJ7f commented on src/proof/er/er_proof.cpp:
```diff
@@ -126,12 +127,12 @@ ErProof ErProof::fromBinaryDratProof(const ClauseUseRecord& usedClauses,
-  dratStream.close();
   tracecheckStream.close();
 
+  unlink(formulaFilename);
```
Nit: `remove()` is standard C while `unlink()` is POSIX, so it is a bit better to use `remove()` here.

alex-ozdemir replied:
Sounds good! Will do.

alex-ozdemir commented:
Okay, @4tXJ7f , I think I've addressed your latest round of comments. Sorry for the whole-file format. I got a little sloppy there.

Paging @mpreiner. The regression tests that I added are currently skipped because our CI builds are not configured with `drat2er`. Could we add `drat2er` to their configuration?
